### PR TITLE
RefreshIndicator can be shown when dragging from non-zero scroll position

### LINKED ISSUE #71304: RefreshIndicator can not pull out when the scrollable widget has a non-zero scroll position
### Bug Reproduce Screenshot

I have tested the Android native behavior, it can be pulled out from the non-zero scroll position.
### Sample Code
```dart
import 'dart:io';

import 'package:flutter/material.dart';

void main() {
  runApp(BugReport());
}

class BugReport extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            title: Text('Demo'),
          ),
          body: Stack(
            children: [
              RefreshIndicator(
                displacement: 100,
                onRefresh: () {
                  return Future.delayed(
                      Duration(seconds: 2), () => print('call onRefresh'));
                },
                child: ListView.builder(
//                    reverse: true,
                    padding: const EdgeInsets.only(top: 0),
                    itemCount: 15,
                    itemBuilder: (BuildContext context, int index) {
                      return Container(
                        height: 50,
                        color: Colors.amber[(index % 2 + 1) * 400],
                        child: Center(child: Text('Entry $index')),
                      );
                    }),
              ),
//              Container(
//                  width: double.infinity,
//                  height: 75,
//                  color: Colors.red,
//                  child: Text(
//                      'Sticky Header that will animate in next implementation')),
            ],
          ),
        ),
      ),
    );
  }
}

```

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

This change fixes that `RefreshIndicator ` can not pull out when the scrollable widget has a non-zero scroll position.

I have tested the Android native behavior, it can be pulled out from the non-zero scroll position.

This change also triggers a `flutter_gallery` APP bug, see changed files for more information.

## Related Issues

Fixes #71304

## Tests

I added the following tests:

See files.

### CODE REVIEW

File: dev/integration_tests/flutter_gallery/test/smoke_test.dart [hunk 1]
```diff
@@ -84,7 +84,12 @@ Future<void> smokeDemo(WidgetTester tester, GalleryDemo demo) async {
   // Scroll the demo around a bit more.
   await tester.flingFrom(const Offset(400.0, 300.0), const Offset(0.0, 400.0), 1000.0);
   await tester.pump();
-  await tester.pump(const Duration(milliseconds: 400));
+  if (demo.title == 'Pull to refresh') {
+    // This fling gesture will trigger 'OverscrollDemo' [RefreshIndicator.onRefresh].
```
Comment (xu-baolin) [inline dev/integration_tests/flutter_gallery/test/smoke_test.dart:88]: https://github.com/flutter/flutter/blob/183f0e797a3bf8aa1b35b650150f7522d5d10377/dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart#L28-L42

File: dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart [hunk 2]
```diff
@@ -27,8 +27,10 @@ class OverscrollDemoState extends State<OverscrollDemo> {
 
   Future<void> _handleRefresh() {
     final Completer<void> completer = Completer<void>();
-    Timer(const Duration(seconds: 3), () { completer.complete(); });
+    Timer(const Duration(seconds: 3), () => completer.complete());
     return completer.future.then((_) {
+      if (!mounted)
```
Comment (xu-baolin) [inline dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart:32]: There is a bug, trigger as following steps:
1. Open the `Pull to refresh` page
2. Pull out the `RefreshIndicator` by dragging
3. Page back before the `_handleRefresh` future complete.
4. Bug triggered. (use `context` after disposal assert)

**Comment by xu-baolin** [comment]
@Piinks @chunhtai Hi,
would you guys take a look at this patch? Thanks :)

File: dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart (hunk 2 above)
Comment (chunhtai) [reply dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart:32]: This seems to exist before the change?

**Review by chunhtai — APPROVED** [review]
LGTM, nice fix!

File: dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart (hunk 2 above)
Comment (xu-baolin) [reply dev/integration_tests/flutter_gallery/lib/demo/material/overscroll_demo.dart:32]: yes,  the bug exist before this change and 
this change can trigger the bug.

**Comment by lazylazyllama** [comment]
There is a catch. Since this PR is merged, there is no overscroll effect anymore. 
If you scroll up with momentum you should get an overscroll effect, but it does not happen anymore.

The overscroll effect should only be disabled, if you reached the top and continue to swipe.

**Comment by chunhtai** [comment]
@lazylazyllama That sounds like a regression, can you file a new issue for it?

**Comment by xu-baolin** [comment]
@lazylazyllama Thanks for the catching.
Yes, you are right, I also test the Android native behavior, the RefreshIndicator should not be pulled out when the scrollable widget rolls to the top by inertia( scroll up with the momentum, release finger before reaching the top edge) and over scroll happens.

I will rework the change!

**Pull request merged**